File: app/src/main/java/com/example/myapplication/TrendsFragment.java
package com.example.myapplication;

import android.app.DatePickerDialog;
import android.content.Context;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class TrendsFragment extends Fragment {

    private enum Mode { CONSECUTIVE_MONTHS, SIX_MONTHS, CONSECUTIVE_YEARS, CUSTOM }

    private Mode mode = Mode.SIX_MONTHS;
    private Calendar customFrom = null;
    private Calendar customTo = null;
    private int consecutiveCount = 6; // months or years

    private final SimpleDateFormat displayFmt = new SimpleDateFormat("MMM yy", Locale.getDefault());
    private final SimpleDateFormat yearFmt = new SimpleDateFormat("yyyy", Locale.getDefault());
    private final SimpleDateFormat pickFmt = new SimpleDateFormat("d MMM yyyy", Locale.getDefault());

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_trends, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Spinner spinner = view.findViewById(R.id.spinnerRange);
        List<String> options = Arrays.asList("3 Months", "6 Months", "12 Months", "3 Years", "5 Years", "Year-over-Year", "Custom");
        spinner.setAdapter(new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_dropdown_item, options));
        spinner.setSelection(1); // default: 6 months

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View v, int pos, long id) {
                LinearLayout customLayout = view.findViewById(R.id.layoutCustomRange);
                switch (pos) {
                    case 0:
                        mode = Mode.CONSECUTIVE_MONTHS;
                        consecutiveCount = 3;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 1:
                        mode = Mode.CONSECUTIVE_MONTHS;
                        consecutiveCount = 6;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 2:
                        mode = Mode.CONSECUTIVE_MONTHS;
                        consecutiveCount = 12;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 3:
                        mode = Mode.CONSECUTIVE_YEARS;
                        consecutiveCount = 3;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 4:
                        mode = Mode.CONSECUTIVE_YEARS;
                        consecutiveCount = 5;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 5:
                        mode = Mode.SIX_MONTHS;
                        customLayout.setVisibility(View.GONE);
                        load(view);
                        break;
                    case 6:
                        mode = Mode.CUSTOM;
                        customLayout.setVisibility(View.VISIBLE);
                        break;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });

        // Custom date pickers
        view.findViewById(R.id.tvCustomFrom).setOnClickListener(v -> pickDate(view, true));
        view.findViewById(R.id.tvCustomTo).setOnClickListener(v -> pickDate(view, false));
        view.findViewById(R.id.btnApplyCustom).setOnClickListener(v -> {
            if (customFrom != null && customTo != null) {
                load(view);
            } else {
                Toast.makeText(requireContext(), "Select both dates", Toast.LENGTH_SHORT).show();
            }
        });

        load(view);
    }

    @Override
    public void onResume() {
        super.onResume();
        View view = getView();
        if (view != null) {
            load(view);
        }
    }

    public void reload() {
        if (isAdded()) {
            View view = getView();
            if (view != null) {
                load(view);
            }
        }
    }

    private void pickDate(View view, boolean isFrom) {
        if (!isVisible()) return;
        Calendar cal = Calendar.getInstance();
        new DatePickerDialog(requireContext(), (view2, y, m, d) -> {
            Calendar picked = Calendar.getInstance();
            picked.set(y, m, d);
            if (isFrom) {
                customFrom = picked;
                ((TextView) view.findViewById(R.id.tvCustomFrom)).setText("From: " + pickFmt.format(picked.getTime()));
            } else {
                customTo = picked;
                ((TextView) view.findViewById(R.id.tvCustomTo)).setText("To: " + pickFmt.format(picked.getTime()));
            }
        }, cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH)).show();
    }

    private void load(View view) {
        DatabaseHelper db = DatabaseHelper.getInstance(requireContext());
        List<BarChartView.MonthData> months = new ArrayList<>();
        String title = "";

        Calendar cal = Calendar.getInstance();

        if (mode == Mode.CONSECUTIVE_MONTHS) {
            title = consecutiveCount + "-Month Trends";
            for (int offset = consecutiveCount - 1; offset >= 0; offset--) {
                Calendar c = (Calendar) cal.clone();
                c.add(Calendar.MONTH, -offset);
                c.set(Calendar.DAY_OF_MONTH, 1);
                c.set(Calendar.HOUR_OF_DAY, 0); c.set(Calendar.MINUTE, 0); c.set(Calendar.SECOND, 0); c.set(Calendar.MILLISECOND, 0);
                long start = c.getTimeInMillis();
                
                c.set(Calendar.DAY_OF_MONTH, c.getActualMaximum(Calendar.DAY_OF_MONTH));
                c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59); c.set(Calendar.MILLISECOND, 999);
                long end = c.getTimeInMillis();

                List<Transaction> txns = db.getTransactions(start, end);
                float income = 0;
                float expense = 0;
                for (Transaction t : txns) {
                    if ("income".equalsIgnoreCase(t.getType())) income += t.getAmount();
                    else if ("expense".equalsIgnoreCase(t.getType())) expense += t.getAmount();
                }
                months.add(new BarChartView.MonthData(displayFmt.format(new Date(start)), income, expense));
            }
        } else if (mode == Mode.CONSECUTIVE_YEARS) {
            title = consecutiveCount + "-Year Trends";
            for (int offset = consecutiveCount - 1; offset >= 0; offset--) {
                Calendar c = (Calendar) cal.clone();
                c.add(Calendar.YEAR, -offset);
                c.set(Calendar.DAY_OF_YEAR, 1);
                c.set(Calendar.HOUR_OF_DAY, 0); c.set(Calendar.MINUTE, 0); c.set(Calendar.SECOND, 0); c.set(Calendar.MILLISECOND, 0);
                long start = c.getTimeInMillis();

                c.set(Calendar.MONTH, 11);
                c.set(Calendar.DAY_OF_MONTH, 31);
                c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59); c.set(Calendar.MILLISECOND, 999);
                long end = c.getTimeInMillis();

                List<Transaction> txns = db.getTransactions(start, end);
                float income = 0;
                float expense = 0;
                for (Transaction t : txns) {
                    if ("income".equalsIgnoreCase(t.getType())) income += t.getAmount();
                    else if ("expense".equalsIgnoreCase(t.getType())) expense += t.getAmount();
                }
                months.add(new BarChartView.MonthData(yearFmt.format(new Date(start)), income, expense));
            }
        } else if (mode == Mode.SIX_MONTHS) {
            title = "Year-over-Year (6 mo)";
            int thisYear = cal.get(Calendar.YEAR);
            for (int offset = 5; offset >= 0; offset--) {
                for (int yr : new int[]{thisYear - 1, thisYear}) {
                    Calendar c = (Calendar) cal.clone();
                    c.set(Calendar.YEAR, yr);
                    c.add(Calendar.MONTH, -offset);
                    c.set(Calendar.DAY_OF_MONTH, 1);
                    c.set(Calendar.HOUR_OF_DAY, 0); c.set(Calendar.MINUTE, 0); c.set(Calendar.SECOND, 0); c.set(Calendar.MILLISECOND, 0);
                    long start = c.getTimeInMillis();

                    c.set(Calendar.DAY_OF_MONTH, c.getActualMaximum(Calendar.DAY_OF_MONTH));
                    c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59); c.set(Calendar.MILLISECOND, 999);
                    long end = c.getTimeInMillis();

                    List<Transaction> txns = db.getTransactions(start, end);
                    float income = 0;
                    float expense = 0;
                    for (Transaction t : txns) {
                        if ("income".equalsIgnoreCase(t.getType())) income += t.getAmount();
                        else if ("expense".equalsIgnoreCase(t.getType())) expense += t.getAmount();
                    }
                    String yrLabel = SimpleDateFormat.getDateInstance().format(new Date(start)); // fallback
                    String labelStr = new SimpleDateFormat("MMM", Locale.getDefault()).format(new Date(start)) + " '" + String.valueOf(yr).substring(2);
                    months.add(new BarChartView.MonthData(labelStr, income, expense));
                }
            }
        } else if (mode == Mode.CUSTOM) {
            if (customFrom == null || customTo == null) return;
            title = pickFmt.format(customFrom.getTime()) + " – " + pickFmt.format(customTo.getTime());

            Calendar cur = (Calendar) customFrom.clone();
            cur.set(Calendar.DAY_OF_MONTH, 1);
            cur.set(Calendar.HOUR_OF_DAY, 0); cur.set(Calendar.MINUTE, 0); cur.set(Calendar.SECOND, 0); cur.set(Calendar.MILLISECOND, 0);

            Calendar toMonth = (Calendar) customTo.clone();
            toMonth.set(Calendar.DAY_OF_MONTH, 1);
            toMonth.set(Calendar.HOUR_OF_DAY, 0); toMonth.set(Calendar.MINUTE, 0); toMonth.set(Calendar.SECOND, 0); toMonth.set(Calendar.MILLISECOND, 0);

            while (!cur.after(toMonth)) {
                long start = cur.getTimeInMillis();
                Calendar endCal = (Calendar) cur.clone();
                endCal.set(Calendar.DAY_OF_MONTH, cur.getActualMaximum(Calendar.DAY_OF_MONTH));
                endCal.set(Calendar.HOUR_OF_DAY, 23); endCal.set(Calendar.MINUTE, 59); endCal.set(Calendar.SECOND, 59); endCal.set(Calendar.MILLISECOND, 999);
                long end = endCal.getTimeInMillis();

                List<Transaction> txns = db.getTransactions(start, end);
                float income = 0;
                float expense = 0;
                for (Transaction t : txns) {
                    if ("income".equalsIgnoreCase(t.getType())) income += t.getAmount();
                    else if ("expense".equalsIgnoreCase(t.getType())) expense += t.getAmount();
                }
                months.add(new BarChartView.MonthData(displayFmt.format(new Date(start)), income, expense));
                cur.add(Calendar.MONTH, 1);
            }
        }

        ((TextView) view.findViewById(R.id.tvTrendsTitle)).setText(title);
        BarChartView chart = view.findViewById(R.id.barChart);
        chart.setData(months);

        float monthlyLimit = requireContext().getSharedPreferences("budget", Context.MODE_PRIVATE).getFloat("monthly_limit", 0f);
        chart.setBudgetLimit(
            (mode == Mode.CONSECUTIVE_MONTHS || mode == Mode.CUSTOM || mode == Mode.SIX_MONTHS) ? monthlyLimit : 0f
        );

        LinearLayout ll = view.findViewById(R.id.llMonthRows);
        ll.removeAllViews();
        ll.addView(makeRow("Period", "Income", "Expense", "Net", true, 0f));
        ll.addView(divider());
        for (BarChartView.MonthData m : months) {
            float net = m.getIncome() - m.getExpense();
            ll.addView(makeRow(
                m.getLabel(),
                "₹" + (int) m.getIncome(),
                "₹" + (int) m.getExpense(),
                (net >= 0 ? "+" : "") + "₹" + (int) net,
                false,
                net
            ));
        }
    }

    private LinearLayout makeRow(String c1, String c2, String c3, String c4, boolean isHeader, float net) {
        Context ctx = requireContext();
        LinearLayout row = new LinearLayout(ctx);
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setPadding(0, 10, 0, 10);

        row.addView(createTextView(ctx, c1, 1.6f, Gravity.START, isHeader, null, net));
        row.addView(createTextView(ctx, c2, 1.0f, Gravity.END, isHeader, ContextCompat.getColor(ctx, R.color.color_income), net));
        row.addView(createTextView(ctx, c3, 1.0f, Gravity.END, isHeader, ContextCompat.getColor(ctx, R.color.color_expense), net));
        row.addView(createTextView(ctx, c4, 1.0f, Gravity.END, isHeader, ContextCompat.getColor(ctx, net >= 0 ? R.color.color_income : R.color.color_expense), net));

        return row;
    }

    private TextView createTextView(Context ctx, String text, float weight, int align, boolean isHeader, Integer customColor, float net) {
        TextView tv = new TextView(ctx);
        tv.setText(text);
        tv.setTextSize(13f);
        tv.setGravity(align);
        if (customColor != null && !isHeader) {
            tv.setTextColor(customColor);
        } else {
            tv.setTextColor(ContextCompat.getColor(ctx, isHeader ? R.color.text_primary : R.color.text_secondary));
        }
        if (isHeader) {
            tv.setTypeface(null, Typeface.BOLD);
        }
        tv.setLayoutParams(new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, weight));
        return tv;
    }

    private View divider() {
        View v = new View(requireContext());
        v.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1));
        v.setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.divider));
        return v;
    }
}

File: app/src/main/java/com/example/myapplication/TrendsFragment.kt
package com.example.myapplication

import android.app.DatePickerDialog
import android.graphics.Typeface
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import java.text.SimpleDateFormat
import java.util.*

class TrendsFragment : Fragment() {

    private enum class Mode { CONSECUTIVE_MONTHS, SIX_MONTHS, CONSECUTIVE_YEARS, CUSTOM }

    private var mode = Mode.SIX_MONTHS
    private var customFrom: Calendar? = null
    private var customTo: Calendar? = null
    // For consecutive modes: how many units back from now
    private var consecutiveCount = 6  // months or years

    private val displayFmt = SimpleDateFormat("MMM yy", Locale.getDefault())
    private val yearFmt    = SimpleDateFormat("yyyy", Locale.getDefault())
    private val pickFmt    = SimpleDateFormat("d MMM yyyy", Locale.getDefault())

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
        inflater.inflate(R.layout.fragment_trends, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val spinner = view.findViewById<Spinner>(R.id.spinnerRange)
        val options = listOf("3 Months", "6 Months", "12 Months", "3 Years", "5 Years", "Year-over-Year", "Custom")
        spinner.adapter = ArrayAdapter(requireContext(), android.R.layout.simple_spinner_dropdown_item, options)
        spinner.setSelection(1) // default: 6 months

        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(p: AdapterView<*>?) {}
            override fun onItemSelected(p: AdapterView<*>?, v: View?, pos: Int, id: Long) {
                val customLayout = view.findViewById<LinearLayout>(R.id.layoutCustomRange)
                when (pos) {
                    0 -> { mode = Mode.CONSECUTIVE_MONTHS; consecutiveCount = 3;  customLayout.visibility = View.GONE; load(view) }
                    1 -> { mode = Mode.CONSECUTIVE_MONTHS; consecutiveCount = 6;  customLayout.visibility = View.GONE; load(view) }
                    2 -> { mode = Mode.CONSECUTIVE_MONTHS; consecutiveCount = 12; customLayout.visibility = View.GONE; load(view) }
                    3 -> { mode = Mode.CONSECUTIVE_YEARS;  consecutiveCount = 3;  customLayout.visibility = View.GONE; load(view) }
                    4 -> { mode = Mode.CONSECUTIVE_YEARS;  consecutiveCount = 5;  customLayout.visibility = View.GONE; load(view) }
                    5 -> { mode = Mode.SIX_MONTHS;         customLayout.visibility = View.GONE; load(view) }
                    6 -> { mode = Mode.CUSTOM; customLayout.visibility = View.VISIBLE }
                }
            }
        }

        // Custom date pickers
        view.findViewById<TextView>(R.id.tvCustomFrom).setOnClickListener { pickDate(view, isFrom = true) }
        view.findViewById<TextView>(R.id.tvCustomTo).setOnClickListener   { pickDate(view, isFrom = false) }
        view.findViewById<Button>(R.id.btnApplyCustom).setOnClickListener {
            if (customFrom != null && customTo != null) load(view)
            else Toast.makeText(requireContext(), "Select both dates", Toast.LENGTH_SHORT).show()
        }

        load(view)
    }

    override fun onResume() { super.onResume(); view?.let { load(it) } }

    private fun pickDate(view: View, isFrom: Boolean) {
        if (!isVisible) return
        val cal = Calendar.getInstance()
        DatePickerDialog(requireContext(), { _, y, m, d ->
            val picked = Calendar.getInstance().apply { set(y, m, d) }
            if (isFrom) {
                customFrom = picked
                view.findViewById<TextView>(R.id.tvCustomFrom).text = "From: ${pickFmt.format(picked.time)}"
            } else {
                customTo = picked
                view.findViewById<TextView>(R.id.tvCustomTo).text = "To: ${pickFmt.format(picked.time)}"
            }
        }, cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH)).show()
    }

    private fun load(view: View) {
        val db = DatabaseHelper(requireContext())
        val months: List<BarChartView.MonthData>
        val title: String

        when (mode) {
            Mode.CONSECUTIVE_MONTHS -> {
                title = "$consecutiveCount-Month Trends"
                val cal = Calendar.getInstance()
                months = (consecutiveCount - 1 downTo 0).map { offset ->
                    val c = (cal.clone() as Calendar).apply {
                        add(Calendar.MONTH, -offset)
                        set(Calendar.DAY_OF_MONTH, 1)
                        set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0)
                    }
                    val start = c.timeInMillis
                    c.set(Calendar.DAY_OF_MONTH, c.getActualMaximum(Calendar.DAY_OF_MONTH))
                    c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59)
                    val txns = db.getTransactions(start, c.timeInMillis)
                    BarChartView.MonthData(
                        displayFmt.format(Date(start)),
                        txns.filter { it.type == "income" }.sumOf { it.amount }.toFloat(),
                        txns.filter { it.type == "expense" }.sumOf { it.amount }.toFloat()
                    )
                }
            }
            Mode.CONSECUTIVE_YEARS -> {
                title = "$consecutiveCount-Year Trends"
                val cal = Calendar.getInstance()
                months = (consecutiveCount - 1 downTo 0).map { offset ->
                    val c = (cal.clone() as Calendar).apply {
                        add(Calendar.YEAR, -offset)
                        set(Calendar.DAY_OF_YEAR, 1)
                        set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0)
                    }
                    val start = c.timeInMillis
                    c.set(Calendar.MONTH, 11); c.set(Calendar.DAY_OF_MONTH, 31)
                    c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59)
                    val txns = db.getTransactions(start, c.timeInMillis)
                    BarChartView.MonthData(
                        yearFmt.format(Date(start)),
                        txns.filter { it.type == "income" }.sumOf { it.amount }.toFloat(),
                        txns.filter { it.type == "expense" }.sumOf { it.amount }.toFloat()
                    )
                }
            }
            Mode.SIX_MONTHS -> {
                // Year-over-year: same 6 months this year vs last year
                title = "Year-over-Year (6 mo)"
                val cal = Calendar.getInstance()
                val thisYear = cal.get(Calendar.YEAR)
                months = (5 downTo 0).flatMap { offset ->
                    listOf(thisYear - 1, thisYear).map { yr ->
                        val c = (cal.clone() as Calendar).apply {
                            set(Calendar.YEAR, yr)
                            add(Calendar.MONTH, -offset)
                            set(Calendar.DAY_OF_MONTH, 1)
                            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0)
                        }
                        val start = c.timeInMillis
                        c.set(Calendar.DAY_OF_MONTH, c.getActualMaximum(Calendar.DAY_OF_MONTH))
                        c.set(Calendar.HOUR_OF_DAY, 23); c.set(Calendar.MINUTE, 59); c.set(Calendar.SECOND, 59)
                        val txns = db.getTransactions(start, c.timeInMillis)
                        val label = SimpleDateFormat("MMM", Locale.getDefault()).format(Date(start)) +
                                " '${yr.toString().takeLast(2)}"
                        BarChartView.MonthData(
                            label,
                            txns.filter { it.type == "income" }.sumOf { it.amount }.toFloat(),
                            txns.filter { it.type == "expense" }.sumOf { it.amount }.toFloat()
                        )
                    }
                }
            }
            Mode.CUSTOM -> {
                val from = customFrom ?: return
                val to   = customTo   ?: return
                title = "${pickFmt.format(from.time)} – ${pickFmt.format(to.time)}"
                // Build month-by-month between from and to
                val result = mutableListOf<BarChartView.MonthData>()
                val cur = (from.clone() as Calendar).apply {
                    set(Calendar.DAY_OF_MONTH, 1)
                    set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0)
                }
                val toMonth = (to.clone() as Calendar).apply {
                    set(Calendar.DAY_OF_MONTH, 1)
                    set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0)
                }
                while (!cur.after(toMonth)) {
                    val start = cur.timeInMillis
                    val end = (cur.clone() as Calendar).apply {
                        set(Calendar.DAY_OF_MONTH, getActualMaximum(Calendar.DAY_OF_MONTH))
                        set(Calendar.HOUR_OF_DAY, 23); set(Calendar.MINUTE, 59); set(Calendar.SECOND, 59)
                    }.timeInMillis
                    val txns = db.getTransactions(start, end)
                    result.add(BarChartView.MonthData(
                        displayFmt.format(Date(start)),
                        txns.filter { it.type == "income" }.sumOf { it.amount }.toFloat(),
                        txns.filter { it.type == "expense" }.sumOf { it.amount }.toFloat()
                    ))
                    cur.add(Calendar.MONTH, 1)
                }
                months = result
            }
        }

        view.findViewById<TextView>(R.id.tvTrendsTitle).text = title
        view.findViewById<BarChartView>(R.id.barChart).setData(months)

        val ll = view.findViewById<LinearLayout>(R.id.llMonthRows)
        ll.removeAllViews()
        ll.addView(makeRow("Period", "Income", "Expense", "Net", isHeader = true))
        ll.addView(divider())
        months.forEach { m ->
            val net = m.income - m.expense
            ll.addView(makeRow(
                m.label,
                "₹${m.income.toInt()}",
                "₹${m.expense.toInt()}",
                (if (net >= 0) "+" else "") + "₹${net.toInt()}",
                isHeader = false, net = net
            ))
        }
    }

    private fun makeRow(c1: String, c2: String, c3: String, c4: String,
                        isHeader: Boolean, net: Float = 0f): LinearLayout {
        val ctx = requireContext()
        return LinearLayout(ctx).apply {
            orientation = LinearLayout.HORIZONTAL
            setPadding(0, 10, 0, 10)
            fun tv(text: String, weight: Float, align: Int = Gravity.START, color: Int? = null) =
                TextView(ctx).apply {
                    this.text = text
                    textSize = 13f
                    gravity = align
                    setTextColor(color ?: ContextCompat.getColor(ctx, if (isHeader) R.color.text_primary else R.color.text_secondary))
                    if (isHeader) setTypeface(null, Typeface.BOLD)
                    layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, weight)
                }
            addView(tv(c1, 1.6f))
            addView(tv(c2, 1f, Gravity.END, if (!isHeader) ContextCompat.getColor(ctx, R.color.color_income) else null))
            addView(tv(c3, 1f, Gravity.END, if (!isHeader) ContextCompat.getColor(ctx, R.color.color_expense) else null))
            addView(tv(c4, 1f, Gravity.END, if (!isHeader) ContextCompat.getColor(ctx,
                if (net >= 0) R.color.color_income else R.color.color_expense) else null))
        }
    }

    private fun divider() = View(requireContext()).apply {
        layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1)
        setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.divider))
    }
}
